### PR TITLE
GH-39958: [Python][CI] Remove upper pin on pytest

### LINKED ISSUE #39957: [Python][CI] S3 fixtures fail with Pytest 8.0.0
### Describe the bug, including details regarding any error messages, version, and platform.

S3 fixtures from out test suite fail with the new `pytest=8.0.0` release with:

```
==================================== ERRORS ====================================
__________ ERROR at setup of test_parquet_file_with_filesystem[True] ___________

s3_server = {'connection': ('localhost', 37805, 'arrow', 'apachearrow'), 'process': <Popen: returncode: None args: ['minio', '--compat', 'server', '--quiet', '-...>, 'tempdir': local('/tmp/pytest-of-root/pytest-0')}

    @pytest.fixture
    def s3_example_fs(s3_server):
        from pyarrow.fs import FileSystem
    
        host, port, access_key, secret_key = s3_server['connection']
        uri = (
            "s3://{}:{}@mybucket/data.parquet?scheme=http&endpoint_override={}:{}"
            .format(access_key, secret_key, host, port)
        )
        fs, path = FileSystem.from_uri(uri)
    
>       fs.create_dir("mybucket")

opt/conda/envs/arrow/lib/python3.10/site-packages/pyarrow/tests/parquet/conftest.py:88: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyarrow/_fs.pyx:603: in pyarrow._fs.FileSystem.create_dir
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: Bucket 'mybucket' not found. To create buckets, enable the allow_bucket_creation option.
```

see
- https://github.com/ursacomputing/crossbow/actions/runs/7738627959/job/21099839415#step:6:5835
- https://github.com/apache/arrow/pull/39850#issuecomment-1920816023 (failing builds from a PR that temporarily removed the upper pin on pytest)

Pytest currently has an upper pin so our CI is not failing.

### Component(s)

Python

### LINKED ISSUE #39958: [Python][CI] Remove upper pin on pytest
### Describe the bug, including details regarding any error messages, version, and platform.

The latest version of pytest (`8.0.0`) can't be used on our CI as it breaks couple of things. This issue is an umbrella issue for all needed fixes that would enable the removal of the upper pin on pytest:

- [ ] https://github.com/apache/arrow/pull/39850
- [ ] https://github.com/lgpage/pytest-cython/issues/58 (disabling `doctest-cython` on our CI until the issue is solved upstream). CI failure can be found here: https://github.com/apache/arrow/actions/runs/7710390851/job/21013608732#step:6:6571)
- [ ] https://github.com/apache/arrow/issues/39957

### Component(s)

Python

### PR DESCRIPTION
**Pull request by AlenkaF**
### Rationale for this change

The latest version of pytest (`8.0.0`) is breaking our CI:
- S3 fixture from out test suite fails
- `doctest-cython` check fails

### What changes are included in this PR?

- added `allow_bucket_creation=True` to the `s3_example_fs` fixture
- removed the pin on pytest, except for the doc builds

### Are these changes tested?

Yes.

### Are there any user-facing changes?
No

Closes:

- Closes https://github.com/apache/arrow/issues/39958
- Closes https://github.com/apache/arrow/issues/39957

* GitHub Issue: #39958

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit -g python

**Review by kou — APPROVED** [review]
+1 if CI is green

BTW,  could you use our PR template instead of just listing related issue links?

**Comment by AlenkaF** [comment]
> BTW, could you use our PR template instead of just listing related issue links?

Will do, thank you for reminding me!

**Comment by AlenkaF** [comment]
@jorisvandenbossche do you mind giving a quick look at this changes?

**Comment by jorisvandenbossche** [comment]
I would personally prefer keeping doctest-cython running on our CI. Is it urgent to remove the pytest pin? Or we could also leave the pin just for the docs build where the doctests are ran?

**Comment by jorisvandenbossche** [comment]
> added `allow_bucket_creation=True` to the `s3_example_fs` fixture

Do we have any understanding about why pytest 8 needs this change?

**Comment by AlenkaF** [comment]
> I would personally prefer keeping doctest-cython running on our CI. Is it urgent to remove the pytest pin? Or we could also leave the pin just for the docs build where the doctests are ran?

We can also wait with merging this PR till https://github.com/lgpage/pytest-cython/issues/58 is merged. I am OK with any of the proposed options.

> > added `allow_bucket_creation=True` to the `s3_example_fs` fixture
> 
> Do we have any understanding about why pytest 8 needs this change?

No 😊 Will research.

**Comment by AlenkaF** [comment]
I have kept the `doctest-cython` command on our CI (conda-python-docs) and pinned pytest to lower than 8 in this case.

As for the `allow_bucket_creation=True` in the `s3_example_fs` fixture, we have agreed there is no point in spending too much time on researching why it is needed. Looking at the PR that added this argument in other cases where `S3FileSystem` is created from a URI (https://github.com/apache/arrow/pull/13206) it makes sense to also add it to the `s3_example_fs` fixture.

File: docker-compose.yml [hunk 1]
```diff
@@ -1208,15 +1208,13 @@ services:
...
-      # https://github.com/lgpage/pytest-cython/issues/24 is resolved
-      # and a new version that includes the fix is released.
-      PYTEST_ARGS: "--doctest-modules --doctest-cython --disable-warnings"
+      PYTEST_ARGS: "--doctest-modules --doctest-cython"
     volumes: *conda-volumes
     command:
       ["/arrow/ci/scripts/cpp_build.sh /arrow /build &&
         /arrow/ci/scripts/python_build.sh /arrow /build &&
         pip install -e /arrow/dev/archery[numpydoc] &&
+        pip install pytest<8 &&
```
Comment (jorisvandenbossche) [inline docker-compose.yml:1217]: Can you add a comment here why this pin is needed (and maybe link to the upstream pytest-cython issue that needs to be resolved to be able to remove this)

Comment (AlenkaF) [reply docker-compose.yml:1217]: Sure, will do!

But I think this pin is causing CI failures. Not sure what should be changed though. I tried using double quotes (https://github.com/apache/arrow/pull/40487/commits/5ccb996cf036fa60880f9a24e6aa22ab88392c96) but it also resulted in failures: https://github.com/apache/arrow/actions/runs/8323506004/job/22773253818

File: docker-compose.yml [hunk 2]
```diff
@@ -1208,15 +1208,17 @@ services:
...
     volumes: *conda-volumes
+    # pytest is installed with an upper pin of 8.0.0 because
+    # newer version breaks cython doctesting, see:
+    # https://github.com/lgpage/pytest-cython/issues/58
+    # Remove pip install pytest~=7 when upstream issue is resolved
     command:
       ["/arrow/ci/scripts/cpp_build.sh /arrow /build &&
         /arrow/ci/scripts/python_build.sh /arrow /build &&
         pip install -e /arrow/dev/archery[numpydoc] &&
+        pip install pytest~=7 &&
```
Comment (jorisvandenbossche) [inline docker-compose.yml:1221]:
```suggestion
        pip install pytest~=7.4 &&
```

Apparently you can't use `~` with a single major number (https://packaging.python.org/en/latest/specifications/version-specifiers/#compatible-release)

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit -g python

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**